module/synchronization/core.go

```diff
@@ -165,8 +169,11 @@ func (c *Core) ScanPending(final *flow.Header) ([]flow.Range, []flow.Batch) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	// prune before doing any work
-	c.prune(final)
+	// prune if the current height is less than the new height
+	if c.localFinalizedHeight < int(final.Height) {
+		c.prune(final)
+		c.localFinalizedHeight = int(final.Height)
+	}
```

Comment:
is `prune` possibly being called by other method? If yes, it's better to move this check into the prune function.